### PR TITLE
perf: compute selected token logprobs directly

### PR DESCRIPTION
**Pull request by CarlosGomes98**
# What does this PR do ?

Compute the selected logprobs directly instead of subsequently masking. Doesn't materialize the full vocabulary-sized logprob tensor. This reduces memory use and unnecessary computation.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by CarlosGomes98** [comment]
/ok to test [7cfd947](https://github.com/NVIDIA-NeMo/RL/pull/3930/commits/7cfd947515292cd4591b8d1f2bdaa0d261e8298d)

File: tests/unit/distributed/test_model_utils.py [hunk 1]
```diff
@@ -42,6 +43,80 @@
...
+    shard_size = full_logits.shape[-1] // 2
+    for rank in range(2):
+        vocab_start = rank * shard_size
+        vocab_end = vocab_start + shard_size
+        target_mask = (target < vocab_start) | (target >= vocab_end)
+        masked_target = target - vocab_start
+        masked_target[target_mask] = 0
+        remote_selected_logprobs = expected.masked_fill(~target_mask, 0.0)
+
+        def fake_all_reduce(tensor, *, op, group):
```
Comment (CarlosGomes98) [inline tests/unit/distributed/test_model_utils.py:125]: This test raises `RuntimeError` before it ever reaches the assertion. The source makes *two* `ReduceOp.SUM` all-reduces (first on `sum_exp_logits` of shape `[2,5,1]`, then on `selected_logprobs` of shape `[2,5]`), but `fake_all_reduce` can't distinguish them, so it runs `tensor.add_(remote_selected_logprobs)` on `sum_exp_logits` and hits a broadcast error (`5` vs `2`). Also note the `sum_exp` reduction goes through `torch.distributed.all_reduce`, not `torch.distributed.nn.functional.all_reduce`, so that monkeypatch is never exercised. Disambiguate the two SUM reduces, e.g.:

```python
def fake_all_reduce(tensor, *, op, group):
    if op == torch.distributed.ReduceOp.MAX:
        tensor.copy_(global_max)
    elif tensor.dim() == 3:          # sum_exp_logits
        tensor.copy_(global_sum_exp)
    else:                            # selected_logprobs
        tensor.add_(remote_selected_logprobs)
```

With that the test passes (and the `nn.functional.all_reduce` monkeypatch can be dropped as dead).

**Review by CarlosGomes98 — COMMENTED** [review]
From the /review-pr skill, as requested:
Core logprob math verified equivalent to the log-softmax path; one blocking issue in the new TP-shard test.

**Comment by CarlosGomes98** [comment]
/ok to test [7cfd947](https://github.com/NVIDIA-NeMo/RL/pull/3930/commits/7cfd947515292cd4591b8d1f2bdaa0d261e8298d)

**Comment by CarlosGomes98** [comment]
/ok to test [ec226ec](https://github.com/NVIDIA-NeMo/RL/pull/3930/commits/ec226ecf223ec9b357f247dfbfb1db23f337c993)

**Comment by CarlosGomes98** [comment]
/ok to test [7ed8ac3](https://github.com/NVIDIA-NeMo/RL/pull/3930/commits/7ed8ac396865ff44f712885727a1df6e6949e9e1)

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -82,6 +82,54 @@ def _compute_distributed_log_softmax(
...
+        sum_exp_logits,
+        op=torch.distributed.ReduceOp.SUM,
+        group=group,
+    )
+    selected_logprobs = selected_logits - sum_exp_logits.log().squeeze(-1).to(
+        selected_logits.dtype
+    )
+    selected_logprobs[target_mask] = 0.0
+
+    torch.distributed.all_reduce(
```
Comment (yuki-97) [inline nemo_rl/distributed/model_utils.py]: The final SUM all-reduce is inside the helper at [`:125`](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L125), so a chunked caller pays one collective per chunk. Making it optional lets both chunked callers hoist it to a single reduce on the concatenated `[B, S]` — `num_chunks` collectives become 1.

`ChunkedDistributedHiddenStatesToLogprobs.forward` is the second caller this would serve: [`:2388-2400`](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L2388-L2400) is the same full-log-softmax then gather-one-index then zero-mask sequence this PR just replaced, its [backward](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L2460-L2468) recomputes from the saved hidden states so the full-vocab tensor is dead there too, and it already hoists its reduce to [`:2405-2409`](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L2405-L2409). It would pick up the 3-buffers-to-1 saving for free.

**Suggest fix:**

1. Add `reduce_output: bool = True` to the helper, documenting that `False` means the caller owns the final reduce, to save `num_chunks - 1` collectives per call; `ChunkedDistributedLogprob.forward` then passes `False` and reduces once after `torch.cat`.
2. Have `ChunkedDistributedHiddenStatesToLogprobs.forward` call the helper too, also with `reduce_output=False`, replacing its [`:2388-2400`](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L2388-L2400) block — its hoisted reduce at [`:2405-2409`](https://github.com/NVIDIA-NeMo/RL/blob/7ed8ac396865ff44f712885727a1df6e6949e9e1/nemo_rl/distributed/model_utils.py#L2405-L2409) already does the right thing.

```python
def _compute_distributed_selected_logprobs(..., reduce_output: bool = True):
    ...
    selected_logprobs[target_mask] = 0.0
    if reduce_output:
        torch.distributed.all_reduce(
            selected_logprobs, op=torch.distributed.ReduceOp.SUM, group=group
        )
    return selected_logprobs
```

Existing tests are unaffected by the default, but the `False` branch is worth one — asserting the ops seen by a monkeypatched `all_reduce` are `[MAX, SUM]` rather than three, since a values-only assertion passes even if the flag is accepted and ignored.

**Review by yuki-97 — COMMENTED** [review]
@CarlosGomes98 Thanks for the improvement. The improvement itself LGTM — went through it end to end and it holds; notes below so the next reader does not have to redo them.

**Checked:**

- Numerical equivalence — the log-normalizer is a per-row scalar, so gathering before or after subtracting it is the same float32 operation on the same operands. Bitwise identical over randomized shapes and scales, and bitwise equal to the unchunked `DistributedLogprob`.
- In-place safety — `shifted_logits` is a fresh allocation and `torch.gather` copies out before `exp_()` runs, so the caller's tensor (saved for backward) is untouched even when the fp32 cast is a no-op on a non-dense slice. Same idiom `_compute_distributed_softmax` already uses.
- Collectives — three per chunk before and after, same order, data-independent, so no rank can diverge; the `nn.functional.all_reduce` to plain swap is inert under `@torch.no_grad`.
- Coverage — `test_distributed_logprob.py::test_distributed_logprob_forward_and_backward[2-4]` exercises the only caller over real NCCL at `tp_size=2`, forward and backward, against a `log_softmax` baseline.

**One comment** below, on the last `all_reduce` and whether we can use this helper in another function as well.

File: nemo_rl/distributed/model_utils.py (hunk 2 above)
Comment (CarlosGomes98) [reply nemo_rl/distributed/model_utils.py]: Nice! Added that.

**Review by yuki-97 — APPROVED** [review]
